File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:

    int longestCommonSubsequence(string text1, string text2) {
        int n=text1.size();
        int m=text2.size();
        vector<int> prev(m+1, 0);
        vector<int> curr(m+1, 0);
        for(int i=1;i<n+1;i++){
            for(int j=1;j<m+1;j++){
                if(text1[i-1]==text2[j-1]){
                    curr[j] = 1+prev[j-1];
                }
                else{
                    curr[j] = max(curr[j-1], prev[j]);
                }
            }
            prev=curr;
        }
        return prev[m];
    }
};

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    public int solveLongestCommonSubsequence(int[][] dp, String text1, String text2, int n, int m) {
        if (n == 0 || m == 0) {
            return 0;
        }
        if (dp[n][m] != -1) {
            return dp[n][m];
        }
        if (text1.charAt(n - 1) == text2.charAt(m - 1)) {
            dp[n][m] = 1 + solveLongestCommonSubsequence(dp, text1, text2, n - 1, m - 1);
            return dp[n][m];
        } else {
            dp[n][m] = Math.max(solveLongestCommonSubsequence(dp, text1, text2, n, m - 1),
                                solveLongestCommonSubsequence(dp, text1, text2, n - 1, m));
            return dp[n][m];
        }
    }

    public int longestCommonSubsequence(String text1, String text2) {
        int n = text1.length();
        int m = text2.length();
        int[][] dp = new int[n + 1][m + 1];
        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }
        return solveLongestCommonSubsequence(dp, text1, text2, n, m);
    }
}